core/src/pretty.rs

```diff
@@ -888,19 +881,31 @@ where
                     .group()
                     .append(allocator.intersperse(
                         foralls.iter().map(|i| allocator.as_string(i)),
-                        allocator.space(), //allocator.softline(),
+                        allocator.space(),
                     ))
                     .append(allocator.text("."))
-                    .append(allocator.line())
+                    .append(allocator.softline())
                     .append(curr.to_owned().pretty(allocator))
             }
-            Enum(erows) => erows.pretty(allocator).enclose("[|", "|]"),
-            Record(rrows) => rrows.pretty(allocator).braces(),
+            Enum(erows) => allocator
+                .line()
```

Comment:
should we always add a newline here? Does this mean that something like `[| 'foo |]` will always be rendered with a new line just for `foo` ?

Reply:
`line()` only turns into a newline if the current `group()` doesn't fit on a line. So if `[| 'foo |]` fits on a line, it stays as it is. If it doesn't, the formatting becomes
```
[|
  'foo
|]
```

With `space()` instead of `line()` here it would turn into
```
[| 'foo
|]
```
if the first line just fits into the line length limit.

Reply:
I see, thanks for the explanation. I thought this was softline, but the documentation of `pretty` indeed makes it different: a softline is a `space()` if the _document_ doesn't fit on a page and is a `line()` otherwise, and the latter is a space of the group doesn't fit on a line. Given this definition, is there a reason anymore to turn the `.line()` after the `forall` into a `softline()`? It seems that we don't really care about the notion of `page` in the pretty printer, or do we?

Reply:
I don't think "page" in the `pretty` documentation refers to actual pages. Rather, it seems to really just be about maximum line length. I originally made that change because `Number -> (forall a. a -> String) -> String` was formatted like
```
Number -> (forall a.
a -> String) -> String
```
which looks horrible. `softline()` seemed to work and I didn't think about it too much. But looking back I think the entire pretty printing of arrow types is inconsistend about line breaks. I'll try to play with it a little and see if I can improve it.

Reply:
Ok, this was just a question, I think it's fair to merge like this and then see if we do better in a follow up PR.